### PR TITLE
Adding unit tests and CI

### PR DESCRIPTION
**Pull request by shigupt202**
Resolves https://github.com/Azure/container-scan/issues/3

### CODE REVIEW

File: src/dockleHelper.ts [hunk 1]
```diff
@@ -29,6 +32,25 @@ const TITLE_TITLE = "TITLE";
 const TITLE_SEVERITY = "SEVERITY";
 const TITLE_DESCRIPTION = "DESCRIPTION";
 
+export async function runDockle(): Promise<number> {
+    const docklePath = await getDockle();
+    core.debug(util.format("Dockle executable found at path ", docklePath));
```
Comment (rgsubh) [inline src/dockleHelper.ts:37]: This debug log can be part of  'getDockle()' function.

File: src/dockleHelper.ts [hunk 2]
```diff
@@ -29,6 +32,25 @@ const TITLE_TITLE = "TITLE";
 const TITLE_SEVERITY = "SEVERITY";
 const TITLE_DESCRIPTION = "DESCRIPTION";
 
+export async function runDockle(): Promise<number> {
+    const docklePath = await getDockle();
+    core.debug(util.format("Dockle executable found at path ", docklePath));
+    const dockleEnv = await getDockleEnvVariables();
+    const imageName = inputHelper.imageName;
+
+    const dockleOptions: ExecOptions = {
```
Comment (rgsubh) [inline src/dockleHelper.ts:41]: we can have small function like:
getDocklePath(..);
getDockleArguments(..);
getDokleOptions(..);
and then toorunner.exec();

File: src/dockleHelper.ts (hunk 1 above)
Comment (shigupt202) [reply src/dockleHelper.ts:37]: Sounds good. I'll update that.

File: src/dockleHelper.ts (hunk 2 above)
Comment (shigupt202) [reply src/dockleHelper.ts:41]: Okay. I've put all of it in a single function as most of these are single line statements anyway. But may be we can delegate these to a few functions as you suggested.

File: package.json [hunk 3]
```diff
@@ -5,22 +5,26 @@
   "main": "lib/main.js",
   "scripts": {
     "build": "tsc",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "jest --silent",
+    "test-coverage": "jest --silent --coverage"
```
Comment (ds-ms) [inline package.json:9]: Why `--silent`?

File: package.json [hunk 4]
```diff
@@ -5,22 +5,26 @@
...
-    "table": "^5.4.6",
-    "typed-rest-client": "^1.7.2",
-    "js-yaml": "^3.12.1"
+    "@actions/core": "1.2.3",
+    "@actions/tool-cache": "1.3.1",
+    "download": "7.1.0",
+    "semver": "7.1.3",
+    "table": "5.4.6",
+    "typed-rest-client": "1.7.2",
+    "js-yaml": "3.12.1"
```
Comment (ds-ms) [inline package.json:28]: Are all these versions the latest?

File: package.json [hunk 5]
```diff
@@ -5,22 +5,26 @@
...
   "devDependencies": {
-    "@types/node": "^13.7.7",
-    "@types/js-yaml": "^3.12.1",
-    "typescript": "^3.8.3"
+    "@types/jest": "25.2.3",
+    "@types/js-yaml": "3.12.1",
+    "@types/node": "13.7.7",
+    "jest": "26.0.1",
+    "ts-jest": "26.0.0",
+    "typescript": "3.8.3"
```
Comment (ds-ms) [inline package.json:19]: I see that you're bringing in only 3 new packages.. could you validate if all the newly introduced package references in package-lock file are relevant? 

one way is, remove that file and run an npm install locally

File: __tests__/main.test.ts [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    'GITHUB_SHA': 'sha',
+    'GITHUB_REPOSITORY': 'test_repo',
+    'GITHUB_TOKEN': 'token',
+    'GITHUB_EVENT_PATH': 'test_event.json',
+    'GITHUB_EVENT_NAME': 'pull_request'
+};
+
+Date.now = jest.fn().mockReturnValue(123);
+
+describe('Validate inputs', () => {
```
Comment (ds-ms) [inline __tests__/main.test.ts:24]: I like the way you broke this down into multiple test suites.

Can you have a different file for each of those suites?

File: __tests__/main.test.ts [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+    afterEach(() => {
+        jest.clearAllMocks();
+    });
+
+    test('No allowedlist file is given', () => {
+        const allowedlistHandler = require('../src/allowedlistHandler');
+        expect(allowedlistHandler.init).not.toThrow();
+        expect(mockedFs.existsSync).not.toReturnWith(true);
+        expect(mockedFs.existsSync).toHaveBeenCalledTimes(2);
```
Comment (ds-ms) [inline __tests__/main.test.ts:144]: Why's it called twice?

File: package.json (hunk 4 above)
Comment (shigupt202) [reply package.json:28]: No, I only removed the greater that signs from the versions. Should I replace all these with the latest versions?

File: __tests__/main.test.ts (hunk 7 above)
Comment (shigupt202) [reply __tests__/main.test.ts:144]: Once for testing whether `allowedlist.YAML` exists and then to test for `allowedlist.YML`

File: package.json (hunk 3 above)
Comment (shigupt202) [reply package.json:9]: If `--silent` is not used, all the debug and console logs get printed along with the test results.

File: package.json (hunk 5 above)
Comment (shigupt202) [reply package.json:19]: Removed the package-lock.json and did `npm i`. Looks like only changes are because of removing the greater than sign in versions. I'll check in the new package-lock.json

File: package.json (hunk 3 above)
Comment (ds-ms) [reply package.json:9]: That's okay right? We need something to debug a failed test?

File: __tests__/main.test.ts (hunk 6 above)
Comment (shigupt202) [reply __tests__/main.test.ts:24]: I tried doing that but there seems to be a conflict in using same names for mocked modules:
`Definitions of the following identifiers conflict with those in another file: mockedFs, mockedOs, mockedToolCache, mockedToolRunner, mockedCore, mockedHttpClient, testUtil, env`.
Also, since most of these suites contain only 2-3 tests, I thought it's better to keep them all in the same file instead of different files to avoid duplicate mocks for process environment variables, certain functions and import statements for mocked modules.

Comment (ds-ms) [reply __tests__/main.test.ts:24]: Oh! Have you tried mocking only in the `main.test.ts` file and just having the test suites in the other files?

Comment (shigupt202) [reply __tests__/main.test.ts]: I'm now initializing all the mocks in describe blocks. It worked.
